src/main/java/com/redhat/devtools/lsp4ij/usages/LSPUsageSearcher.java

```diff
@@ -55,84 +60,106 @@ public class LSPUsageSearcher extends CustomUsageSearcher {
+            // Get position where the "Find Usages" has been triggered
+            Position position = getPosition(element, file);
+            if (position == null) {
+                return null;
+            }
+            // Collect textDocument/definition, textDocument/references, etc
+            LSPUsageSupport usageSupport = new LSPUsageSupport(file);
+            LSPUsageSupport.LSPUsageSupportParams params = new LSPUsageSupport.LSPUsageSupportParams(position);
+            CompletableFuture<List<LSPUsagePsiElement>> usagesFuture = usageSupport.getFeatureData(params);
+            try {
+                // Wait for completion of textDocument/definition, textDocument/references, etc
+                waitUntilDone(usagesFuture);
+                if (usagesFuture.isDone()) {
+                    // Show response of textDocument/definition, textDocument/references, etc as usage info.
+                    List<LSPUsagePsiElement> usages = usagesFuture.getNow(null);
+                    if (usages != null) {
+                        for (LSPUsagePsiElement usage : usages) {
+                            processor.process(new UsageInfo2UsageAdapter(new UsageInfo(usage)));
+                        }
                     }
                 }
+            } catch (ProcessCanceledException pce) {
+                throw pce;
+            } catch (Exception e) {
+                LOGGER.error("Error while collection LSP Usages", e);
             }
-        } catch (ProcessCanceledException pce) {
-            throw pce;
-        } catch (Exception e) {
-            LOGGER.error("Error while collection LSP Usages", e);
-        }
 
-        // For completeness' sake, also collect external usages to LSP (pseudo-)elements
-        LSPExternalReferencesFinder.processExternalReferences(
-                file,
-                ReadAction.compute(element::getTextOffset),
-                options.searchScope,
-                reference -> processor.process(ReadAction.compute(() -> new UsageInfo2UsageAdapter(new UsageInfo(reference))))
-        );
+            // For completeness' sake, also collect external usages to LSP (pseudo-)elements
+            LSPExternalReferencesFinder.processExternalReferences(
+                    file,
+                    element.getTextOffset(),
+                    options.searchScope,
+                    reference -> processor.process(new UsageInfo2UsageAdapter(new UsageInfo(reference)))
+            );
+
+
+            return null;
+        }, element);
     }
 
-    @Nullable
-    private static Position getPosition(@NotNull PsiElement element, @NotNull PsiFile psiFile) {
-        if (ApplicationManager.getApplication().isReadAccessAllowed()) {
-            return doGetPosition(element, psiFile);
+    private static void processElementUsagesUnderReadAction(@NotNull ThrowableComputable<CodeVisionState, Throwable> computable,

```

Comment:
Everything in this change makes sense to me except for the addition of this method. Why was this added instead of just wrapping the entire body of `processElementUsages()` in `ReadAction.compute()`?

Reply:
The main reason is because when dumb is processed we need to skip this search. And when we are in non EDT it should be cancellable.

I find this logic code in UJ ultimate repo.